examples/ota_http_client.rs

```diff
@@ -0,0 +1,147 @@
+//! OTA update example.
+//!
+//! In this example, we are checking for updates from a distant server hosting the latest firmware.
+//! The hosting server is supposed to check if an update should be done or not by responding with
+//! a 200 (OK, you should update) or 304 (Not Modified, you are up to date) status code.
+//!
+//! For this example to work, you need a OTA ready partition table.
+//! You need at least 2 OTA partitions (see https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/kconfig.html#config-partition-table-type)
+//!
+//! The most common starting point for OTA is the "Factory app, two OTA definitions" layout (https://github.com/espressif/esp-idf/blob/master/components/partition_table/partitions_two_ota.csv).
+//!
+//! To use a custom partition table, download the CSV file of your choice (or create a custom one)
+//! and use either use the `--partition-table` option of `espflash`, or set this option in your `espflash.toml` file.
+//!
+//! After a successful OTA update, you will need to reset the `otadata` partition. Otherwise, the ESP
+//! will continue to boot on the second partition, while `cargo run` is flashing the first one by default.
+//! To reset the `otadata` partition, add `--erase-parts otadata` to the runner command in `.cargo/config.toml`.
+
+use anyhow::Context;
+use embedded_svc::http::client::Client as HttpClient;
+use esp_idf_svc::http::client::{EspHttpConnection, Method};
+use esp_idf_svc::io;
+use esp_idf_svc::log::EspLogger;
+use esp_idf_svc::wifi::{BlockingWifi, ClientConfiguration, Configuration, EspWifi};
+use esp_idf_svc::{eventloop::EspSystemEventLoop, nvs::EspDefaultNvsPartition};
+use esp_idf_svc::{
+    hal::peripherals::Peripherals,
+    ota::{EspOta, SlotState},
+};
+
+use log::{error, info};
+
+const VERSION: &str = "1.0.0"; // You can pull this from an environment variable at build time using env! macro.
+const OTA_FIRMWARE_URI: &str = "http://your.domain/path/to/firmware";
+
+const SSID: &str = env!("WIFI_SSID");
+const PASSWORD: &str = env!("WIFI_PASS");
+
+mod http_status {
+    pub const OK: u16 = 200;
+    pub const NOT_MODIFIED: u16 = 304;
+}
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_svc::sys::link_patches();
+    EspLogger::initialize_default();
+
+    // Setup Wifi
+
+    let peripherals = Peripherals::take()?;
+    let sys_loop = EspSystemEventLoop::take()?;
+    let nvs = EspDefaultNvsPartition::take()?;
+
+    let mut wifi = BlockingWifi::wrap(
+        EspWifi::new(peripherals.modem, sys_loop.clone(), Some(nvs))?,
+        sys_loop,
+    )?;
+
+    connect_wifi(&mut wifi)?;
+
+    let mut client = HttpClient::wrap(EspHttpConnection::new(&Default::default())?);
+
+    check_for_updates(&mut client)?;
+
+    check_valid_state();
+
+    Ok(())
+}
+
+fn check_valid_state() -> anyhow::Result<()> {
```

Comment:
Perhaps document that this functionality needs `CONFIG_BOOTLOADER_APP_ROLLBACK_ENABLE=y` which is not the case by default?

Reply:
How this have to be set ? Is there a file somewhere for the sep-idf related configuration?

Reply:
`sdkconfig.defaults` in the root of your binary crate. You should absolutely have this file there anyway.
Also, since this is a bootloader setting, you _must_ flash the custom bootloader that is anyway built as part of building your app crate, or else the check_valid_state fn would do essentially nothing, as the standard bootloader flashed by `espflash` is unlikely to have this setting enabled (it is not enabled by default).

Reply:
From what I've understood of this section: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/ota.html#app-rollback
the default behavior of the bootloader is to select a not invalid slot. So it can be either Valid or Pending Validation. Which means that if the slot is explicitly marked as invalid, it will not boot.

I will test this to be sure about it.